alicat/driver.py

```diff
@@ -0,0 +1,484 @@
+             'C2H4', 'i-C2H10', 'Kr', 'Xe', 'SF6', 'C-25', 'C-10',
+             'C-8', 'C-2', 'C-75', 'A-75', 'A-25', 'A1025', 'Star29',
+             'P-5']
+
+    def __init__(self, address='/dev/ttyUSB0', unit='A', **kwargs):
+        """Connect this driver with the appropriate USB / serial port.
+
+        Args:
+            address: The serial port or TCP address:port. Default '/dev/ttyUSB0'.
+            unit: The Alicat-specified unit ID, A-Z. Default 'A'.
+        """
+        if address.startswith('/dev') or address.startswith('COM'):  # serial
+            self.hw: Client = SerialClient(address=address, **kwargs)
+        else:
+            self.hw = TcpClient(address=address, **kwargs)
+
+        self.unit = unit
+        self.keys = ['pressure', 'temperature', 'volumetric_flow', 'mass_flow',
+                     'setpoint', 'gas']
+        self.open = True
+        self.firmware: Union[str, None] = None
+
+    async def __aenter__(self, *args):
+        """Provide async enter to context manager."""
+        return self
+
+    async def __aexit__(self, *args):
+        """Provide async exit to context manager."""
+        return
+
+    @classmethod
+    async def is_connected(cls, port, unit='A') -> bool:
+        """Return True if the specified port is connected to this device.
+
+        This class can be used to automatically identify ports with connected
+        Alicats. Iterate through all connected interfaces, and use this to
+        test. Ports that come back True should be valid addresses.
+
+        Note that this distinguishes between `FlowController` and `FlowMeter`.
+        """
+        is_device = False
+        try:
+            device = cls(port, unit)
+            try:
+                c = await device.get()
+                if cls.__name__ == 'FlowMeter':
+                    assert c
+                    assert 'setpoint' not in device.keys
+                elif cls.__name__ == 'FlowController':
+                    assert c
+                    assert 'setpoint' in device.keys
+                else:
+                    raise NotImplementedError('Must be meter or controller.')
+                is_device = True
+            finally:
+                await device.close()
+        except Exception:
+            pass
+        return is_device
+
+    def _test_controller_open(self) -> None:
+        """Raise an IOError if the FlowMeter has been closed.
+
+        Does nothing if the meter is open and good for read/write
+        otherwise raises an IOError. This only checks if the meter
+        has been closed by the FlowMeter.close method.
+        """
+        if not self.open:
+            raise OSError(f"The FlowMeter with unit ID {self.unit} and "
+                           "port {self.hw.address} is not open")
+
+    def _is_float(self, msg):
+        try:
+            float(msg)
+            return True
+        except ValueError:
+            return False
```

Comment:
Pedantic, but I'd move outside of the class.

Reply:
1fc791397e226cac6dfa50770fe833c9d232c0fa